File: ios/Classes/BvYouTubeVideoView.swift
//
//  YoutuyerView.swift
//  Pods-Runner
//
//  Created by Agnaramon Boris-Carnot on 19/07/2019.
//

import Foundation
import Flutter
import youtube_ios_player_helper
import SnapKit

enum PlayerState: String {
    case UNKNOWN = "UNKNOWN"
    case UNSTARTED = "UNSTARTED"
    case ENDED = "ENDED"
    case PLAYING = "PLAYING"
    case PAUSED = "PAUSED"
    case BUFFERING = "BUFFERING"
    case VIDEO_CUED = "VIDEO_CUED"
}

enum PlayerError: String {
    case UNKNOWN = "UNKNOWN"
    case INVALID_PARAMETER_IN_REQUEST = "INVALID_PARAMETER_IN_REQUEST"
    case HTML_5_PLAYER = "HTML_5_PLAYER"
    case VIDEO_NOT_FOUND = "VIDEO_NOT_FOUND"
    case VIDEO_NOT_PLAYABLE_IN_EMBEDDED_PLAYER = "VIDEO_NOT_PLAYABLE_IN_EMBEDDED_PLAYER"
}

class BvYouTubeVideoView: NSObject, FlutterPlatformView{

    private let frame: CGRect
    private let viewId: Int64
    private let params: [String: Any]
    private let playerView: UIView
    private let channel: FlutterMethodChannel
    private var isPlayerReady = false
    private var player: YTPlayerView!
    private let messenger: FlutterBinaryMessenger

    let playerVars: [String: Any] = [
        "origin" : "https://www.youtube.com",
        "rel": 0,
        "autoplay": 0
    ]

    init(messenger: FlutterBinaryMessenger,
         frame: CGRect,
         viewId: Int64,
         params: [String: Any]?
         ) {
        self.messenger = messenger
        self.frame = frame
        self.viewId = viewId
        self.params = params!
        playerView = UIView(frame: frame)
        channel = FlutterMethodChannel(
            name: "bv_youtube_video",
            binaryMessenger: messenger
        )
        super.init()
        self.initPlayer()
        channel.setMethodCallHandler { [weak self] (call, result) in
            guard let `self` = self else { return }
            `self`.handle(call, result: result)
        }
    }

    func view() -> UIView {
        return playerView
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialization":
            result(nil)
        case "loadOrCueVideo":
            print("loadOrCueVideo is called")
            let params = call.arguments as! Dictionary<String, Any>
            let videoId = params["videoId"] as! String
            let startSeconds = params["startSeconds"] as? Double ?? 0.0
            loadOrCueVideo(videoId: videoId, startSeconds: startSeconds)
            result(nil)
        case "play":
            print("play is called")
            if (self.isPlayerReady) {
                self.player.playVideo()
            }
            result(nil)
        case "pause":
            print("pause is called")
            if (self.isPlayerReady) {
                self.player.pauseVideo()
            }
            result(nil)
        case "seekTo":
            print("seekTo is called")
            if (self.isPlayerReady) {
                let second = call.arguments as! Double
                self.player.seek(toSeconds: Float(Int(second)), allowSeekAhead: true)
            }
            result(nil)
        case "mute":
            print("mute is called")
            //self.player.mute()
            result(nil)
        case "unMute":
            print("mute is called")
            //self.player.unMute()
            result(nil)
        case "setVolume":
            print("setVolume is called")
            result(nil)
        case "scaleMode":
            /*let scaleMode = call.arguments as! Int
            self.changeScaleMode(scaleMode: scaleMode)
            UIView.animate(withDuration: 0.3) { [weak self] in
                guard let `self` = self else { return }
                self.playerView.layoutIfNeeded()
            }*/
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func loadOrCueVideo(videoId: String, startSeconds: Double = 0.0) {
        if (!self.isPlayerReady) {
            return
        }
        player.load(withVideoId: videoId, playerVars: playerVars)
    }

    func initPlayer(){

        let videoId = params["videoId"] as? String
        let showUI = params["showUI"] as! Bool
        let scaleMode = params["scale_mode"] as? Int ?? 0

        self.playerView.backgroundColor = .yellow
        self.player = YTPlayerView()
        self.playerView.addSubview(self.player)

        self.player.snp.makeConstraints{ make in
            make.edges.equalToSuperview()
        }

        player.delegate = self
        player.load(withVideoId: videoId!, playerVars: playerVars)

        //player.cueVideo(byId: "2kHnFYzYC-0", startSeconds: 0.0, suggestedQuality: .medium)

    }

    private func onStateChange(state: YTPlayerState) {
        var customState: PlayerState!
        switch state {
        case .cued:
            customState = .VIDEO_CUED
        case .ended:
            customState = .ENDED
        case .playing:
            customState = .PLAYING
        case .paused:
            customState = .PAUSED
        case .buffering:
            customState = .BUFFERING
        case .unstarted:
            customState = .UNSTARTED
        case .unknown:
            break
        }
        print("state = \(state)")
        channel.invokeMethod("onStateChange", arguments: customState.rawValue)
    }

}

extension BvYouTubeVideoView: YTPlayerViewDelegate{
    func playerViewDidBecomeReady(_ playerView: YTPlayerView) {
        print(#function)
        let startSeconds = (params["startSeconds"] as? Double ?? 0.0)
        let autoStart = (params["autoStart"] as? Bool ?? false)
        if (startSeconds > 0 || autoStart ) {
            player.seek(toSeconds: Float(Int(startSeconds)), allowSeekAhead: true)
        }
        self.isPlayerReady = true
        channel.invokeMethod("onReady", arguments: nil)
    }

    func playerView(_ playerView: YTPlayerView, didPlayTime playTime: Float) {
        print("\(#function):\(playTime)")
        channel.invokeMethod("onCurrentSecond", arguments: playTime)
    }

    func playerView(_ playerView: YTPlayerView, didChangeTo state: YTPlayerState) {
        print("\(#function):\(state)")
        if (state == YTPlayerState.playing) {
            player.duration { result, error in
                self.channel.invokeMethod("onVideoDuration", arguments: result)
            }
        }
        self.onStateChange(state: state)
    }

    func playerView(_ playerView: YTPlayerView, didChangeTo quality: YTPlaybackQuality) {
         print("\(#function):\(quality)")
    }
}

File: android/src/main/kotlin/com/bottlevin/bv_youtube_video/BvYouTubeVideoView.kt
package com.bottlevin.bv_youtube_video

import android.app.Activity
import android.app.Application
import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.Gravity
import android.view.View
import android.widget.FrameLayout
import androidx.lifecycle.Lifecycle
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.views.YouTubePlayerView
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.platform.PlatformView
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.PlayerConstants
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.YouTubePlayer
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.listeners.AbstractYouTubePlayerListener
import java.util.concurrent.atomic.AtomicReference

class BvYouTubeVideoView(
        private val context: Context?,
        id: Int,
        private val params: Map<String, Any?>,
        private val state: AtomicReference<Lifecycle.Event>,
        private val messenger: BinaryMessenger
) :
        PlatformView,
        MethodChannel.MethodCallHandler,
        ActivityAware {

    private val TAG = "BvYouTubeVideoView"

    private lateinit var youtubePlayerView: YouTubePlayerView
    private lateinit var container: FLTPlayerView
    private val view: FrameLayout
    private var youtubePlayer: YouTubePlayer? = null
    private var activity: Activity? = null
    private val methodChannel: MethodChannel

    init {
        val mode = params["scale_mode"] as? Int ?: 0
        view = createView(mode = mode)
        changeScaleMode(mode = mode)
        methodChannel = MethodChannel(messenger, "bv_youtube_video")
        methodChannel.setMethodCallHandler(this)
        initYouTubePlayerView()
    }

    private fun createView(mode: Int = 0): FrameLayout {
        val videoMode = BvYouTubeVideoScaleMode.values().find { it.mode == mode }
        container = FLTPlayerView(context!!).apply {
            setVideoScaleMode(mode = videoMode!!)
        }
        container.apply {
            youtubePlayerView = YouTubePlayerView(context)
            val layoutParams = FrameLayout.LayoutParams(
                    FrameLayout.LayoutParams.MATCH_PARENT,
                    FrameLayout.LayoutParams.WRAP_CONTENT
            )
            addView(youtubePlayerView, layoutParams)
        }
        return FrameLayout(context!!).apply {
            val layoutParams = FrameLayout.LayoutParams(
                    FrameLayout.LayoutParams.MATCH_PARENT,
                    FrameLayout.LayoutParams.MATCH_PARENT
            )
            layoutParams.gravity = Gravity.CENTER
            addView(container, layoutParams)
        }
    }

    private fun initYouTubePlayerView() {
        Log.d(TAG, "params = $params")
        val videoId = params["videoId"] as? String
        val startSeconds = (params["startSeconds"] as Double).toFloat()
        val showUI = params["showUI"] as Boolean
        /*val controller = youtubePlayerView.getPlayerUiController()
        controller.showYouTubeButton(false)
        controller.showFullscreenButton(false)
        if (!showUI) {
            controller.showUi(false)
            controller.showVideoTitle(false)
        }*/
        youtubePlayerView.addYouTubePlayerListener(object : AbstractYouTubePlayerListener() {
            override fun onReady(youTubePlayer: YouTubePlayer) {
                youtubePlayer = youTubePlayer
                methodChannel.invokeMethod("onReady", null)
                if (videoId != null) {
                    loadOrCueVideo(videoId, startSeconds)
                }
            }

            override fun onStateChange(
                    youTubePlayer: YouTubePlayer,
                    state: PlayerConstants.PlayerState
            ) {
                onStateChange(state)
            }

            override fun onError(youTubePlayer: YouTubePlayer, error: PlayerConstants.PlayerError) {
                onError(error)
            }

            override fun onVideoDuration(youTubePlayer: YouTubePlayer, duration: Float) {
                methodChannel.invokeMethod("onVideoDuration", duration)
            }

            override fun onCurrentSecond(youTubePlayer: YouTubePlayer, second: Float) {
                methodChannel.invokeMethod("onCurrentSecond", second)
            }
        })
    }

    override fun getView(): View {
        return view
    }

    override fun dispose() {
        youtubePlayerView.release()
//        registrar.activity().application.unregisterActivityLifecycleCallbacks(this)
    }

    override fun onMethodCall(methodCall: MethodCall, result: MethodChannel.Result) {
        when (methodCall.method) {
            "initialization" -> result.success(null)
            "loadOrCueVideo" -> loadOrCueVideo(methodCall, result)
            "play" -> play(result)
            "pause" -> pause(result)
            "seekTo" -> seekTo(methodCall, result)
            "setVolume" -> setVolume(methodCall, result)
            "mute" -> {
                youtubePlayer?.setVolume(0)
                result.success(null)
            }
            "unMute" -> {
                youtubePlayer?.setVolume(100)
                result.success(null)
            }
            "scaleMode" -> {
                changeScaleMode(methodCall.arguments as Int)
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    private fun loadOrCueVideo(methodCall: MethodCall, result: MethodChannel.Result) {
        val params = methodCall.arguments as HashMap<String, *>
        val videoId = params["videoId"] as String
        val startSeconds = (params["startSeconds"] as? Double ?: 0.0).toFloat()
        loadOrCueVideo(videoId, startSeconds)
        result.success(null)
    }

    private fun loadOrCueVideo(videoId: String, startSeconds: Float) {
        val canLoad = state.get() == Lifecycle.Event.ON_RESUME
        if (canLoad)
            youtubePlayer?.loadVideo(videoId, startSeconds)
        else
            youtubePlayer?.cueVideo(videoId, startSeconds)
    }

    private fun pause(result: MethodChannel.Result) {
        youtubePlayer?.pause()
        result.success(null)
    }

    private fun play(result: MethodChannel.Result) {
        youtubePlayer?.play()
        result.success(null)
    }

    private fun seekTo(methodCall: MethodCall, result: MethodChannel.Result) {
        val time = (methodCall.arguments as Double).toFloat()
        youtubePlayer?.seekTo(time)
        result.success(null)
    }

    /**
     * @param volumePercent Integer between 0 and 100
     */
    private fun setVolume(methodCall: MethodCall, result: MethodChannel.Result) {
        val volumePercent = methodCall.arguments as Int
        youtubePlayer?.setVolume(volumePercent)
        result.success(null)
    }

    private fun changeScaleMode(mode: Int) {
        val videoMode = BvYouTubeVideoScaleMode.values().find { it.mode == mode }
        container.setVideoScaleMode(videoMode!!)
        val playerHeight = when (videoMode) {
            BvYouTubeVideoScaleMode.NONE -> FrameLayout.LayoutParams.MATCH_PARENT
            BvYouTubeVideoScaleMode.FIT_WIDTH -> FrameLayout.LayoutParams.WRAP_CONTENT
            BvYouTubeVideoScaleMode.FIT_HEIGHT -> FrameLayout.LayoutParams.WRAP_CONTENT
        }
        val videoParams = youtubePlayerView.layoutParams.apply {
            width = FrameLayout.LayoutParams.MATCH_PARENT
            height = playerHeight
        }
        youtubePlayerView.layoutParams = videoParams
        val containerWidth = when (videoMode) {
            BvYouTubeVideoScaleMode.NONE -> FrameLayout.LayoutParams.MATCH_PARENT
            BvYouTubeVideoScaleMode.FIT_WIDTH -> FrameLayout.LayoutParams.MATCH_PARENT
            BvYouTubeVideoScaleMode.FIT_HEIGHT -> FrameLayout.LayoutParams.WRAP_CONTENT
        }
        val containerHeight = when (videoMode) {
            BvYouTubeVideoScaleMode.NONE -> FrameLayout.LayoutParams.MATCH_PARENT
            BvYouTubeVideoScaleMode.FIT_WIDTH -> FrameLayout.LayoutParams.WRAP_CONTENT
            BvYouTubeVideoScaleMode.FIT_HEIGHT -> FrameLayout.LayoutParams.MATCH_PARENT
        }
        val containerGravity = when (videoMode) {
            BvYouTubeVideoScaleMode.NONE -> Gravity.CENTER
            BvYouTubeVideoScaleMode.FIT_WIDTH -> Gravity.CENTER_HORIZONTAL
            BvYouTubeVideoScaleMode.FIT_HEIGHT -> Gravity.CENTER_HORIZONTAL
        }
        val containerParams = (container.layoutParams as FrameLayout.LayoutParams).apply {
            width = containerWidth
            height = containerHeight
            gravity = containerGravity
        }
        container.layoutParams = containerParams
        view.requestLayout()
    }

    private fun onStateChange(state: PlayerConstants.PlayerState) {
        Log.d(TAG, "state = $state")
        val customState = when (state) {
            PlayerConstants.PlayerState.VIDEO_CUED -> PlayerState.VIDEO_CUED.value
            PlayerConstants.PlayerState.UNSTARTED -> PlayerState.UNSTARTED.value
            PlayerConstants.PlayerState.ENDED -> PlayerState.ENDED.value
            PlayerConstants.PlayerState.PLAYING -> PlayerState.PLAYING.value
            PlayerConstants.PlayerState.PAUSED -> PlayerState.PAUSED.value
            PlayerConstants.PlayerState.BUFFERING -> PlayerState.BUFFERING.value
            else -> PlayerState.UNKNOWN.value
        }
        methodChannel.invokeMethod("onStateChange", customState)
    }

    private fun onError(error: PlayerConstants.PlayerError) {
        Log.d(TAG, "error = ${error.name}")
        val customError = when (error) {
            PlayerConstants.PlayerError.INVALID_PARAMETER_IN_REQUEST -> PlayerError.INVALID_PARAMETER_IN_REQUEST.value
            PlayerConstants.PlayerError.VIDEO_NOT_PLAYABLE_IN_EMBEDDED_PLAYER -> PlayerError.VIDEO_NOT_PLAYABLE_IN_EMBEDDED_PLAYER.value
            PlayerConstants.PlayerError.HTML_5_PLAYER -> PlayerError.HTML_5_PLAYER.value
            PlayerConstants.PlayerError.VIDEO_NOT_FOUND -> PlayerError.VIDEO_NOT_FOUND.value
            else -> PlayerError.UNKNOWN.value
        }
        methodChannel.invokeMethod("onError", customError)
    }


    //--
    // ActivityAware
    //--

    override fun onDetachedFromActivity() {
        this.activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        this.activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity()
    }
}